### ISSUE TITLE
Ray Core: Ray dashboard agent memory usage is high

### ISSUE THREAD

**Issue by tchordia** [ray-project/ray#41040]
### What happened + What you expected to happen

Endpoints Cluster is showing very high memory usage for ray dashboard agent. 

This is pushing memory utilization on the head node all the way up to 90% consistently, and also seems to be causing high cpu utilization. I am worried that this will cause a head node crash which would take down the cluster.

### Versions / Dependencies

2.7.1

### Reproduction script

Not sure what an easy repro is here.

### Issue Severity

None

**Comment by rkooo567** [comment]
@rynewang @jjyao instead of waiting for them to repro, can we actively repro it ourselves? Writing a similar computation pattern workload and see if we can repro in a long running cluster?

**Comment by anyscalesam** [comment]
Discussed - let's at least root cause to know the why of this first. cc @rynewang to take point on this.

**Comment by rkooo567** [comment]
This issue is a collection of multiple issues. 

1. When there are lots of custom metrics with high cardinality, it can increase the agent memory usage.
2. Dashboard agent may have memory leak when cleaning up metrics.

1 is mitigated from the application side already, and they are planning to test soon.
2. is something I tried reproducing. I basically created a script which creates a lot of custom metrics.

```python3
import ray
from ray.util.metrics import Histogram
import time

ray.init()

DEFAULT_LATENCY_BUCKET_MS = [
    i for i in range(1, 1000)
]

@ray.remote
class Actor:
    def __init__(self):
        self.hists = [
            Histogram(
                    f"{i}",
                    description=("my histogram "),
                    boundaries=DEFAULT_LATENCY_BUCKET_MS,
                    tag_keys=())
            for i in range(100)
        ]

    def record(self):
        for hist in self.hists:
            for i in DEFAULT_LATENCY_BUCKET_MS:
                hist.observe(i)

while True:
    actors = [Actor.remote() for _ in range(40)]
    ray.get([actor.record.remote() for actor in actors])
    time.sleep(5)
    del actors
```

And run ray agent with very small metrics export interval (by default, 10 seconds, and I used 10ms).

Investigation
-------------
From OSX and Linux, I observed memory leak-looking symptoms initially, but it turns out not to be a memory leak.

From htop. it shows the high resident size initially, but it eventually shrinked. For OSX, resident size didn't decrease, but not every allocator free the memory immediately. My guess is OSX allocator doesn't free actively although memory is freed.

I verified heap size eventually shrinks for both OS. See the screenshots below.

After that, I ran memray profiler to verify leaks. I verified heap size eventually shrinks for both OS. See the screenshots below.

OSX
<img width="1728" alt="Screenshot 2023-12-12 at 11 14 18 PM" src="https://github.com/ray-project/ray/assets/18510752/1f87c82d-73de-4f1b-9fca-7c3e509248f4">

Linux 
<img width="1719" alt="Screenshot 2023-12-12 at 11 14 39 PM" src="https://github.com/ray-project/ray/assets/18510752/11ed02c4-855f-426b-957d-40c535a82bef">

Note that heap size is still 200MB (which is bigger than expected). It was due to Python allocator (see details here https://bloomberg.github.io/memray/python_allocators.html#). When I tracked the python allocator / deallocator using --trace-python-allocator, I verified the code doesn't have any leak and deallocators are properly called. 

Conclusion
-----------
There's no leak from existing dashboard agent code when metrics are deleted. Thus I will remove the release blocker. We can close the issue once cardinality reduction from Ray serve solves endpoint problem. 

Follow up
----------
We still need a couple of follow up for metrics. This will be more of P1 in Core team's future tasks, but we don't currently have bandwidth to handle them. 

- Need actual memory leak testing
- When cardinality is very high, we should overwrite the tag to reduce cardinality (instead of exploding the dashboard agent).
- We should report volume of metrics generated.

**Comment by rkooo567** [comment]
Keep it P0 until @alexeykudinkin verifies his fix

**Comment by rkooo567** [comment]
cc @jjyao @rynewang

**Comment by rynewang** [comment]
I wonder if we can try this [1]: when we invoke gc.collect(), also invoke a `malloc_trim`. I am not sure this would work or not. We can do an experiment for it


https://www.softwareatscale.dev/p/run-python-servers-more-efficiently

**Comment by rkooo567** [comment]
Actually you can also try PYTHONMALLOC=malloc, and I also verified no leaks with this option

**Comment by rynewang** [comment]
I mean, since in production we have the need to ask pymalloc to release memory to the OS we may need some tricks like malloc_trim.

**Comment by alexeykudinkin** [comment]
This is a likely caused by a combination of 

 - (Addressed in 2.9) Serve producing new time-series for every request: https://github.com/ray-project/ray/issues/41282
 - (Addressed in 2.9) Metrics not being delivered to agent from workers: https://github.com/ray-project/ray/issues/40033

We've validated both of these were addressed in 2.9, so closing this one
